packages/state-history/src/lib/entities-state-history.ts

```diff
@@ -0,0 +1,414 @@
+import {
+  Store,
+  OrArray,
+  coerceArray,
+  isUndefined,
+  filterNil,
+  StoreValue,
+} from '@ngneat/elf';
+import { Subscription, pairwise, filter, map } from 'rxjs';
+import {
+  EntitiesState,
+  getIdType,
+  EntitiesRef,
+  getEntityType,
+  DefaultEntitiesRef,
+  upsertEntities,
+  getEntitiesIds,
+  selectEntity,
+  selectEntities,
+} from '@ngneat/elf-entities';
+
+export interface EntitiesStateHistoryOptions<E extends EntitiesRef> {
+  maxAge: number;
+
+  entitiesRef: E | null;
+
+  entityIds?: OrArray<getIdType<EntitiesState<E>, E>>;
+  // comparatorFn: (prev, current) => isEqual(prev, current) === false
+  comparatorFn: (
+    prevState: Array<getEntityType<EntitiesState<E>, E>>,
+    currentState: Array<getEntityType<EntitiesState<E>, E>>
+  ) => boolean;
+}
+
+type EntityHistory<E extends EntitiesRef> = {
+  past: Array<getEntityType<EntitiesState<E>, E>>;
+  present: getEntityType<EntitiesState<E>, E> | null;
+  future: Array<getEntityType<EntitiesState<E>, E>>;
+};
+
+export class EntitiesStateHistory<
+  T extends Store<any, StoreValue<any> & EntitiesState<E>>,
+  E extends EntitiesRef,
+  S extends StoreValue<T> = StoreValue<T>
+> {
+  private entitiesHistory = new Map<
+    getIdType<EntitiesState<E>, E>,
+    EntityHistory<E>
+  >();
+  private entitiesChanges = new Map<
+    getIdType<EntitiesState<E>, E>,
+    Subscription
+  >();
+  private pausedEntitiesChanges = new Set<getIdType<EntitiesState<E>, E>>();
+
+  private skipUpdate = false;
+  private subscription: Subscription | undefined;
+
+  private mergedOptions: EntitiesStateHistoryOptions<E>;
+
+  constructor(
+    protected store: T,
+    private options: Partial<EntitiesStateHistoryOptions<E>> = {}
+  ) {
+    this.mergedOptions = {
+      maxAge: 10,
+      entitiesRef: null,
+      comparatorFn: () => true,
+      ...options,
+    };
+    this.activate();
+  }
+
+  activate() {
+    /**
+     * If we want to check specific entities.
+     */
+    if (this.mergedOptions.entityIds) {
+      coerceArray(this.mergedOptions.entityIds).forEach((id) => {
+        this.subscribeToEntityChanges(id);
+      });
+    } else {
+      this.subscription = this.store
+        .pipe(
+          selectEntities(this.entitiesRef),
+          map(() => this.store.query(getEntitiesIds(this.entitiesRef))),
```

Comment:
why do you need the `getEntitiesIds`? you're taking care of this in the first condition. Seems not relevant.

Reply:
The first condition is for passed specific ids that we want to track. This one is for tracking all entities but if I do `Object.keys` on `selectEntities` response I'll get keys as strings and it's wrong because I usu `Map`s.